src/ngx_http_lua_util.c

```diff
@@ -421,6 +421,11 @@ ngx_http_lua_send_header_if_needed(ngx_http_request_t *r,
             dd("sending headers");
             rc = ngx_http_send_header(r);
             ctx->header_sent = 1;
+
+            if (r->filter_finalize) {
+                ngx_http_set_ctx(r, ctx, ngx_http_lua_module);
```

Comment:
It's also wrong to just keep running the original location's lua code while the newly entered location's lua code may have already been running.